2021/day16/day16.java

```diff
@@ -0,0 +1,132 @@
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Scanner;
+
+public class day16 {
+    public static class Packet {
+        Long value;
+        int length;
+        int versionSum;
+        String binData;
+    }
+
+    public static String readInput(String filename) {
+        StringBuilder result = new StringBuilder();
+        Scanner sc = null;
+        try {
+            sc = new Scanner(new File(filename));
+            String s = sc.next();
+            for (char c : s.toCharArray()) {
+                result.append(hexCharToBinString(c));
+            }
+        } catch (Exception e) {
+            System.out.println("Something went wrong " + e);
+        } finally {
+            sc.close();
+        }
+        return result.toString();
+    }
+
+    public static String hexCharToBinString(char c){
+        int decNum = Integer.parseInt(String.valueOf(c), 16);
+        String binString = String.format("%4s", Integer.toBinaryString(decNum));
+        return binString.replace(' ', '0');
+    }
+
+    public static Long calculateTotalSubpacketValue(int packetType, ArrayList<Long> values) {
+        switch (packetType) {
+            case 0:
+                Long sum = 0L;
+                for (Long num : values) {
+                    sum += num;
+                }
+                return sum;
+            case 1:
+                Long product = 1L;
+                for (Long num : values) {
+                    product *= num;
+                }
+                return product;
+            case 2:
+                return Collections.min(values);
+            case 3:
+                return Collections.max(values);
+            case 5:
+                return values.get(0) > values.get(1) ? 1L : 0L;
+            case 6:
+                return values.get(0) < values.get(1) ? 1L : 0L;
+            case 7:
+                return values.get(0).equals(values.get(1)) ? 1L : 0L;
+            default:
+                System.out.println("Error in switch statement");
+                return -1L;
+        }
+    }
+
+    public static Packet calculateLiteralValue(Packet packet) {
+        int i = 6;
+        StringBuilder literalBinValue = new StringBuilder();
+        do{
+            literalBinValue.append(packet.binData.substring(i + 1, i + 5));
+            i += 5;
+        } while(packet.binData.charAt(i - 5) == '1');
+        packet.binData = packet.binData.substring(i);
+        packet.value = Long.parseLong(literalBinValue.toString(), 2);
+        packet.length = i;
+        return packet;
+    }
+
+    public static Packet processSubpacketsByLength(Packet packet, int packetTypeID) {
+        int subLength = Integer.parseInt(packet.binData.substring(7, 22), 2);
+        int packetLength = 22;
+        packet.binData = packet.binData.substring(22);
```

Comment:
If you already initialise a magic `22` (which - fair enough - is hard to avoid as that's part of the problem statement), you may as well use it in the lines above and below.